[Extensions] Add gift card tracks

ThomazFB opened the pull request:
Plan: pe5pgL-3CT-p2#comment-3025

<!-- Id number of the GitHub issue this PR addresses. -->

## Why
We'd like to understand the gift card usage in the order form after it's released. 

## How

Event | Properties | Triggers
-- | -- | --
order_form_add_gift_card_cta_shown | flow: creation/editing (Order creation or editing) | When the Add Gift Card CTA is shown in the order form (the site has active Gift Cards plugin).
order_form_add_gift_card_cta_tapped | flow: creation/editing | When the user taps the Add Gift Card CTA in the order form.
order_form_gift_card_set | flow: creation/editing, removed: Bool (whether the gift card is removed) | When the user sets the gift card in the order form.
order_creation_success(EXISTING) | use_gift_card: Bool (whether a gift card is used) (NEW) |  
order_creation_failed(EXISTING) | use_gift_card: Bool (whether a gift card is used) (NEW) |  
order_sync_failed(EXISTING) | use_gift_card: Bool (whether a gift card is used) (NEW)

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Store eligible for Gift Cards - create order

- Switch to a store that has the Gift Cards plugin active
- Go to the Orders tab
- Tap `+`
- Add a fee or product with a price so that the order total is non-zero --> after the order total becomes positive, `order_form_add_gift_card_cta_shown` is tracked with `"flow": "creation"`
- Tap `Add Gift Card` --> `order_form_add_gift_card_cta_tapped` is tracked with `"flow": "creation"`
- Enter a gift card code that can't be used --> `order_form_gift_card_set` is tracked with `"removed": false` and `"flow": "creation"`
- Enter the gift card editing view and remove the code entirely and click apply --> `order_form_gift_card_set` is tracked with `"removed": true` and `"flow": "creation"`
- Tap `Add Gift Card` and enter a gift card code that can't be used
- Tap `Create` --> `order_creation_failed` is tracked with `"use_gift_card": true`
- Tap `Add Gift Card` and enter a valid gift card with some remaining balance --> `order_form_gift_card_set` is tracked with `"removed": false` and `"flow": "creation"`. After the order is created, `order_creation_success` is tracked with `"use_gift_card": true`

### Store not eligible for Gift Cards

- Switch to a store that doesn't have the Gift Cards plugin installed/active
- Go to the Orders tab
- Tap on an editable order
- Tap `Edit` --> the Add Gift Card CTA shouldn't be shown, and no `order_form_add_gift_card_cta_shown` is tracked

---
- [ ] @ThomazFB register new events after it's approved
- [ ] @ThomazFB tests that `"use_gift_card": false` for order updates/creations without a gift card

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Add gift card presence track to order creation success

ThomazFB committed:
Add gift card presence track to order creation failure

ThomazFB committed:
Add gift card presence track to order sync failure

ThomazFB committed:
Add gift card new analytics events

ThomazFB committed:
Add all three gift card operations calls

ThomazFB committed:
Connect gift card analytic events with View Model events

ThomazFB committed:
Add gift card tracks unit test coverage

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge branch 'trunk' into issue/ad-gift-card-tracks

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

atorresveiga commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/UnifiedOrderEditViewModelTest.kt:
```diff
@@ -453,13 +453,71 @@ abstract class UnifiedOrderEditViewModelTest : BaseUnitTest() {
     }
 
+    @Test
+    fun `given a standard order creation, when add gift card is clicked, then track expected event`() = testBlocking {
```
Nice!!

atorresveiga commented:
Thanks for working on this, @ThomazFB 🙏.

I found some small issues, but I am not sure if they are related to the merge with the new order details changes:


### 1

> Add a fee or product with a price so that the order total is non-zero 

@ThomazFB  the `order_form_add_gift_card_cta_shown` is tracked when the extension is active without taking into account  the order total


https://github.com/woocommerce/woocommerce-android/assets/18119390/5c203cd4-d3c3-4c61-bf69-fc4f5ee42e43


### 2

>Tap `X` on the gift card row to delete the gift card --> order_form_gift_card_set is tracked with "removed": true and "flow": "creation"

I'm not seeing the `X` button:

<img width="300" src="https://github.com/woocommerce/woocommerce-android/assets/18119390/032facb2-7dcf-44ff-887d-1cc3d621366c" />


To see the event, I needed to tap on edit gift card and entered a blank gift code

```
🔵 Tracked: order_form_gift_card_set, Properties: {"flow":"creation","removed":true,"blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}
```

### 3

I believe I found a bug with the gift card button

Steps to reproduce:
1. Open the orders list
2. Add a new product
3. Tap on add gift card
4. Enter a valid gift card code and press apply
5. Check that the add gift card button is hidden
6. Expand the order total card
7. Tap on edit gift card
8. Press apply again without making any changes
9. **Check that the add gift card button is shown**
10. **Check that the add gift card button is not responsive**

https://github.com/woocommerce/woocommerce-android/assets/18119390/873dcfd4-b70d-4fff-ab9e-4e370d47c86a


Evidence for TC1
```
🔵 Tracked: order_form_add_gift_card_cta_shown, Properties: {"flow":"creation","blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}

🔵 Tracked: order_form_add_gift_card_cta_tapped, Properties: {"flow":"creation","blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}

🔵 Tracked: order_form_gift_card_set, Properties: {"flow":"creation","removed":false,"blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}

🔵 Tracked: order_form_gift_card_set, Properties: {"flow":"creation","removed":false,"blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}

🔵 Tracked: order_creation_failed, Properties: {"error_context":"OrderCreateEditViewModel","error_type":"GENERIC_ERROR","error_description":"Gift card code SXVD-MJKJ-K97C-QT7C not found.","use_gift_card":true,"blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}

🔵 Tracked: order_creation_success, Properties: {"coupons_count":0,"use_gift_card":true,"blog_id":222408976,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"wooexpress-small-bundle-yearly","store_id":"b6468f83-4986-44f1-b4ce-6021649dbf29","is_debug":true,"site_url":"https:\/\/woo-practically-automatic-witch.wpcomstaging.com"}
```

atorresveiga commented:
I don't see the add gift card button for TC2, but the `order_form_add_gift_card_cta_shown` event is tracked. Here is a video: 

https://github.com/woocommerce/woocommerce-android/assets/18119390/db4dc852-54d8-4de0-b60c-db03a2c394b9

atorresveiga commented:
**Store not eligible for Gift Cards** are working as expected 😄

ThomazFB committed:
Merge branch 'trunk' into issue/ad-gift-card-tracks

ThomazFB committed:
Adjust OrderCreateEditViewModel to only trigger gift card CTA shown tracks when the Gift card button is enabled

ThomazFB commented:
Hey @atorresveiga, thanks for your review!

> the order_form_add_gift_card_cta_shown is tracked when the extension is active without taking into account the order total

Good point. I adjusted the tracks trigger in https://github.com/woocommerce/woocommerce-android/pull/10564/commits/2b5703af2999b1a7ce24a8f12dfecfbf931d8eb8. Now we should see it just once, and only when the gift card button is enabled for the first time.

>I'm not seeing the X button:

Sorry, I added the same test flow used in iOS and forgot to adjust the fact that we didn't add the clear button yet. This should be fixed once we have https://github.com/woocommerce/woocommerce-android/pull/10544 merged.

>I believe I found a bug with the gift card button
This bug seems to be related with the current view state control of the Order Creation, I suggest we tackle it in a separate PR, because looks like it will require some rework with the current `AddInfoButtonsStateTransition` controls.

>I don't see the add gift card button for TC2, but the order_form_add_gift_card_cta_shown event is tracked.

Sorry, I forgot to remove the updating scenario. We currently aren't supporting editing the Gift card in the WCAndroid. But I fixed the issue of the track triggering even when the button is unavailable!

ThomazFB committed:
Fix lint issues

atorresveiga reviewed (approved):
Thanks for working o this @ThomazFB 🙏 

LGTM! :shipit:

atorresveiga commented:
> Sorry, I added the same test flow used in iOS and forgot to adjust the fact that we didn't add the clear button yet. This should be fixed once we have https://github.com/woocommerce/woocommerce-android/pull/10544 merged.

np: Maybe we should update the testing instruction of this PR and mention that to remove a Gift Card code, there is no X button and that you need to erase the current code and apply an empty code in the form. WDYT?
